src/libstd/io/mod.rs

```diff
@@ -0,0 +1,928 @@
+// except according to those terms.
+
+//! Traits, helpers, and type definitions for core I/O functionality.
+//!
+//! > **NOTE**: This module is very much a work in progress and is under active
+//! > development. At this time it is still recommended to use the `old_io`
+//! > module while the details of this module shake out.
+
+#![unstable(feature = "io",
+            reason = "this new I/O module is still under active deveopment and \
+                      APIs are subject to tweaks fairly regularly")]
+
+use cmp;
+use unicode::str as core_str;
+use error::Error as StdError;
+use fmt;
+use iter::Iterator;
+use marker::Sized;
+use mem;
+use ops::{Drop, FnOnce};
+use option::Option::{self, Some, None};
+use ptr::PtrExt;
+use result::Result::{Ok, Err};
+use result;
+use slice::{self, SliceExt};
+use string::String;
+use str::{self, StrExt};
+use vec::Vec;
+
+pub use self::buffered::{BufReader, BufWriter, BufStream, LineWriter};
+pub use self::cursor::Cursor;
+pub use self::error::{Result, Error, ErrorKind};
+pub use self::util::{copy, sink, Sink, empty, Empty, repeat, Repeat};
+
+pub mod prelude;
+mod buffered;
+mod cursor;
+mod error;
+mod impls;
+mod util;
+
+const DEFAULT_BUF_SIZE: usize = 64 * 1024;
+
+// Acquires a slice of the vector `v` from its length to its capacity
+// (uninitialized data), reads into it, and then updates the length.
+//
+// This function is leveraged to efficiently read some bytes into a destination
+// vector without extra copying and taking advantage of the space that's already
+// in `v`.
+//
+// The buffer we're passing down, however, is pointing at uninitialized data
+// (the end of a `Vec`), and many operations will be *much* faster if we don't
+// have to zero it out. In order to prevent LLVM from generating an `undef`
+// value when reads happen from this uninitialized memory, we force LLVM to
+// think it's initialized by sending it through a black box. This should prevent
+// actual undefined behavior after optimizations.
+fn with_end_to_cap<F>(v: &mut Vec<u8>, f: F) -> Result<usize>
+    where F: FnOnce(&mut [u8]) -> Result<usize>
+{
+    unsafe {
+        let n = try!(f({
+            let base = v.as_mut_ptr().offset(v.len() as isize);
+            black_box(slice::from_raw_mut_buf(mem::copy_lifetime(v, &base),
+                                              v.capacity() - v.len()))
+        }));
+
+        // If the closure (typically a `read` implementation) reported that it
+        // read a larger number of bytes than the vector actually has, we need
+        // to be sure to clamp the vector to at most its capacity.
+        let new_len = cmp::min(v.capacity(), v.len() + n);
+        v.set_len(new_len);
+        return Ok(n);
+    }
+
+    // Semi-hack used to prevent LLVM from retaining any assumptions about
+    // `dummy` over this function call
+    unsafe fn black_box<T>(dummy: T) -> T {
+        asm!("" : : "r"(&dummy));
+        dummy
+    }
```

Comment:
This is passing a shared reference to an (otherwise immutable) stack slot. The only reason it works is that LLVM doesn't tell the difference - it's like using `const` in C when you clearly want to pretend mutation is happening.
Aside from telling LLVM about its immutability, an immediate `T` with no unsafe interior is safe to promote to a SSA value when debuginfo isn't required, and I have implemented this optimization in all of of my trans refactoring branches.
I highly suggest passing `&mut dummy` to `asm!`, in addition to the `"memory"` clobber suggested by @mahkoh.